extensions/standard-processors/processors/GenerateFlowFile.cpp

```diff
@@ -98,67 +115,32 @@ void GenerateFlowFile::onTrigger(core::ProcessContext *context, core::ProcessSes
     logger_->log_trace("Unique Flow files is configured to be %i", uniqueFlowFile);
   }
 
-  if (uniqueFlowFile) {
-    char *data;
-    data = new char[fileSize];
-    if (!data)
+  for (uint64_t i = 0; i < batchSize; i++) {
+    // For each batch
+    std::shared_ptr<FlowFileRecord> flowFile = std::static_pointer_cast<FlowFileRecord>(session->create());
+    if (!flowFile) {
+      logger_->log_error("Failed to create flowfile!");
       return;
-    uint64_t dataSize = fileSize;
-    GenerateFlowFile::WriteCallback callback(data, dataSize);
-    char *current = data;
-    if (textData) {
-      for (uint64_t i = 0; i < fileSize; i++) {
-        int randValue = random();
-        data[i] = TEXT_CHARS[randValue % TEXT_LEN];
-      }
-    } else {
-      for (uint64_t i = 0; i < fileSize; i += sizeof(int)) {
-        int randValue = random();
-        *(reinterpret_cast<int*>(current)) = randValue;
-        current += sizeof(int);
-      }
     }
-    for (uint64_t i = 0; i < batchSize; i++) {
-      // For each batch
-      std::shared_ptr<FlowFileRecord> flowFile = std::static_pointer_cast<FlowFileRecord>(session->create());
-      if (!flowFile)
-        return;
-      if (fileSize > 0)
+    if (fileSize > 0) {
+      if (uniqueFlowFile) {
+        std::vector<char> data(fileSize);
```

Comment:
If we don't have a unique flow file then we should follow what NiFi does, which is generate the data in onSchedule -- that would obviate the need for locking here and simplify the logic altogether. 

Take a look at 
https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GenerateFlowFile.java#L167

Reply:
Very nice idea, I love it, done!